Editing pass (#26)
Haivision/srt-rfc: draft-sharabayko-mops-srt.md

 | 0xFFFFFFFD | DONE                         |
 | 0xFFFFFFFE | AGREEMENT                    |
 | 0xFFFFFFFF | CONCLUSION                   |

>  | 0x00000000 | WAVEAHAND                    |

Author: This is how it is referred to in Mikolaj's handshake document.

~~~~
{: #keymaterial-extension-structure title="Key Material Extension structure"}

0 ( ): 1 bit. Value: {0}  
: This is a fixed-width field that is a remnant from the header of a previous design (VF).

Version (V): 3 bits. Value: {1}  
: This is a fixed-width field that indicates the SRT version:

- 1: initial version

Packet Type (PT): 4 bits. Value: {2}  
: This is a fixed-width field that indicates the Packet Type:

- 0: Reserved
- 1: MSmsg
- 2: KMmsg 
- 7: Reserved to discriminate MPEG-TS packet (0x47=sync byte)   

Signature (Sign): 16 bits. Value: {0x2029}  

> : This is a fixed-width field that contains the signature ‘HAI‘ encoded as a PnP Vendor ID 

Reviewer: PNPID reference was here

   (in big endian order)

Reserved (Resv): 6 bits. Value: {0}  
: This is a fixed-width field reserved for flag extension or other usage.

> 

Reviewer: SP800-38A reference is missing now

: The sequence number of the last acknowledged data packet +1.

RTT (32 bits):
: RTT value (in microseconds) estimated by the receiver based on the previous ACK-ACKACK 
packet exchange.

RTT variance (32 bits):
: The variance of the RTT estimation (in microseconds).

Available Buffer Size (32 bits):
: Available size of the receiver's buffer (in packets).

Packets Receiving Rate (32 bits):
: The rate at which packets are being received (in packets per second).

Estimated Link Capacity (32 bits):
: Estimated bandwidth of the link (in packets per second).

Receiving Rate (32 bits):
: Estimated receiving rate (in bytes per second).

There are several types of ACK packets:

    - A Full ACK control packet is sent every 10 ms and has all the fields 
      of {{ack-control-packet}}.
    - A Lite ACK control packet includes only the Last Acknowledged Packet Sequence Number 
      field. The Type-specific Information field should be set to 0.
    - A Small ACK includes the fields up to and including the Available Buffer Size field. 
      The Type-specific Information field should be set to 0.

> The sender only acknowledges only the receipt of Full ACK packets (see ACKACK).

Reviewer: "only" is used twice

Author: Good catch!

|0|                            Up to                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|0|                 Lost packet sequence number                 |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ---

Control Type: 
: The type value of a NAK control packet is "3".

Type-specific Information: 
: This field is reserved for future definition.

Control Information Field (CIF):
: A single value or a list of lost packets sequence numbers. See packet sequence number 
coding in {{packet-seq-list-coding}}.

### Shutdown {#ctrl-pkt-shutdown}

Shutdown control packets are used to initiate the closing of an SRT connection.

An SRT SHUTDOWN Control packet is formatted as follows:

 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ---
|1|        Control Type         |           Reserved            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                   Type-specific Information                   |

> +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ SRT Header

Reviewer: This line is 77 characters in width. The conversion to IETF txt format requires 72.
Maybe we can fix it this way:
```
+-+-+-+-+-+-+-+-+SRT Header-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ 
```

|                           Time Stamp                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     Destination Socket ID                     |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ---

> |                              None                             | CIF

Reviewer: CIF = None to reduce the width?

Control Information Field:
: This field must not appear in shutdown control packets.

> 

Reviewer: ```suggestion
```

Control Information Field:
: This field must not appear in ACKACK control packets.

> 

Reviewer: ```suggestion
```

### Message Mode {#transmission-mode-msg}

When the STREAM flag of the handshake Extension Message {#handshake-extension-msg} is set 
to 0, the protocol operates in Message mode, characterized as follows:

    - Every packet has its own Packet Sequence Number.
    - One or several consecutive SRT Data packets can form a message.
    - All the packets belonging to the same message have a similar message number set 
      in the Message Number field.

The first packet of a message has the first bit of the Packet Position Flags ({{data-pkt}})
set to 1. The last packet of the message has the second bit of the Packet Position Flags
set to 1. Thus, a PP equal to "11b" indicates a packet that forms the whole message.
A PP equal to "00b" indicates a packet that belongs to the inner part of the message.

> The concept of the message in SRT comes from UDT ({{I-D.gg-udt}}). In this mode a single 

Reviewer: ```suggestion
The concept of the message in SRT comes from UDT ({{GHG04b}}). In this mode a single 
```

sending instruction passes exactly one piece of data that has boundaries (a message). 
This message may span across multiple UDP packets (and multiple SRT data packets). The 
only size limitation is that it shall fit as a whole in the buffers of the sender and the 
receiver. Although internally all operations (e.g. ACK, NAK) on data packets are performed 
independently, an application MUST send and receive the whole message. Until the message 
is complete (all packets are received) the application will not be allowed to read it.

When the Order Flag of a Data packet is set to 1, this imposes a sequential reading order
on messages. An Order Flag set to 0 allows an application to read messages that are 
already fully available, before any preceding messages that may have some packets missing.

### Live Mode {#transmission-mode-live}

Live mode is a special type of message mode where only data packets
with their PP field set to "11b" are allowed.

> Additionally, TsbPd ({{tsbpd}}) and TL Packet drop ({{tl-pkt-drop}}) mechanisms are used 

Reviewer: Section anchor has been changed.
```suggestion
Additionally, TsbPd ({{tsbpd}}) and TL Packet drop ({{too-late-packet-drop}}) mechanisms are used 
```

Buffer mode is negotiated during the Handshake by setting the STREAM flag
of the handshake Extension Message Flags to 1.

In this mode consecutive packets form one continuous stream that can be read, with 
portions of any size.

## Stream Multiplexing

Multiple SRT sockets may share one UDP socket, and the packets received on this
UDP socket will be correctly dispatched to the SRT socket to which they are currently 
destined. The parties exchange their SRT Socket IDs during the handshake. These IDs are 
then used in the Destination Socket ID field of every control and data packet.

> 

Reviewer: This paragraph was moved into the "SRT Data Transmission and Control" section above (line 831).
So this is a repetition.

Author: Will delete it.

2. After receiving the above message from the peer: CONCLUSION.
3. After receiving the above message from the peer: AGREEMENT.

When a connection process has failed before either party can send the CONCLUSION handshake, 
the Handshake Type field will contain the appropriate error value for the rejected 
connection. See the list of error codes in {{hs-rej-reason}}.

>  | Code | Error            | Description                                    |

Reviewer: Maybe "Error code" is a better name then?

is adjusted to the receiver’s local time (compensating for the time drift or different 
time zones) before releasing the packet to the application. Packets can be withheld by 
the SRT receiver for a configured receiver delay. A higher delay can accommodate a larger 
uniform packet drop rate, or a larger packet burst drop. Packets received after their 
"play time" are dropped if the Too-Late Packet Drop feature is enabled 
(see {{too-late-packet-drop}}).

The packet timestamp (in microseconds) is relative to the SRT connection creation time. 
Packets are inserted based on the sequence number in the header field. The origin time 
(in microseconds) of the packet is already sampled when a packet is first submitted by 
the application to the SRT sender. The TSBPD feature uses this time to stamp the packet 
for first transmission and any subsequent retransmission. This timestamp and the 
configured SRT latency ({{srt-latency}}) control the recovery buffer size and the 
instant that packets are delivered at the destination (the aforementioned "play time" 
which is decided by adding the timestamp to the configured latency).

> It is worth mentioning that the use of the packet sending time to stamp the packets is

Reviewer: This part starting from "Packets are inserted .." is deleted. Was it an intent?

Author: Will replace

The NAK packet in that case will have all the packets that the receiver considers being lost
at the time of sending the Periodic NAK report.

An ACKACK tells the receiver to stop sending the ACK position because the sender already
knows it. Otherwise, ACKs (with outdated information) would continue to be sent regularly.

An ACK serves as a ping, with a corresponding ACKACK pong, to measure RTT. The time it 
takes for an ACK to be sent and an ACKACK to be received is the RTT. Each ACK has a number. 
A corresponding ACKACK has that same number. The receiver keeps a list of all ACKs in a 
queue to match them. Unlike a full ACK, which contains the current RTT and several other 
values in the CIF, a light ACK just contains the sequence number. All control messages 
are sent directly and processed upon reception, but ACKACK processing time is negligible 
(the time this takes is included in the round-trip time).

## Bidirectional Transmission Queues

Once an SRT connection is established, both peers can send data packets simultaneously.

> 

Reviewer: ```suggestion
```

## Round Trip Time Estimation

The round-trip time is estimated during the transmission of SRT data packets
based on the time difference between the ACK packet is sent and the
corresponding ACKACK is received by the data receiver.

> 

Reviewer: ```suggestion
```
